Add wiki to repo (#102)
mesh-adaptation/docs: wiki/Development-Practices.md

The three Mesh Adaptation codes and their documentation repo have a shared [Project board](https://github.com/orgs/mesh-adaptation/projects/2/views/1) for ongoing development work.
The project board categories are:
* **Backlog** - Issues that have been flagged but there are not currently any plans to address them.
* **Priority** - Issues that have been flagged as important and should be addressed as soon as possible.
* **Blocked** - Issues that cannot be progressed until one or more other issues are addressed.
* **In progress** - Issues for which one or more developers are currently working on addressing.
* **Ready for review** - Issues for which there is at least one open Pull Request that has reviewers assigned to it, but which hasn't been reviewed yet.
* **Reviewed with actions** - Issues for which there is at least one open Pull Request that has been reviewed and changes have been requested from the reviewer, but not yet actioned.
* **Discuss** - Issues that should be discussed at an upcoming development meeting.
* **Done** - Issues that have been addressed or are no longer valid.

Notice that all the categories refer to [Issues](#issues).

> [Pull Requests](#pull-requests) should not be put on the project board.

Reviewer: ```suggestion
[Pull Requests](#pull-requests) should not be put on the project board, except if they do not have a linked issue. (See  [below](#issues) for special cases in which this might occur.)
```

> ### Other project boards
> 
> During the course of developing the mesh adaptation codes, it may be useful to set up new project boards, e.g., for sprints or hackathons.

Reviewer: ```suggestion
### Other project board views

By default, you will see the "Overview" view of the project board. This is a somewhat overwhelming representation of all the issues associated with the development of Mesh Adaptation packages. Browse the other "views" of the project board by clicking on the different tabs for more focused representations of issues related to particular pieces of work or specific packages.

The current recommendation is to create new views for new pieces of work, rather than creating new project boards. This helps to keep everything in the same place. Views can be set up in many ways but a common approach is to filter for a particular label, e.g., `docsprint`.
```

## Issues

* If you identify a bug, missing feature, or other issue with one of the codes then raise an Issue for it:
  * New Issue links: <a href="https://github.com/mesh-adaptation/animate/issues/new/choose" target="_blank" rel="noopener noreferrer">Animate</a>, <a href="https://github.com/mesh-adaptation/goalie/issues/new/choose" target="_blank" rel="noopener noreferrer">Goalie</a>, <a href="https://github.com/mesh-adaptation/movement/issues/new/choose" target="_blank" rel="noopener noreferrer">Movement</a>.
* Add labels to your Issue using the right-hand panel, for example identifying whether it relates to a "bug" or an "enhancement". If you think the Issue is of high priority then use the "PRIORITY" label, too.
* Add your new Issue to the [Mesh Adaptation development project board](https://github.com/orgs/mesh-adaptation/projects/2/views/1) and set its status to "Backlog" once the Issue is created. If you used the "PRIORITY" label above then use that status here, too.

> * Set the milestone of the Issue to an appropriate upcoming milestone from the drop-down menu.

Reviewer: ```suggestion
* Set the milestone of the Issue to an appropriate upcoming milestone from the drop-down menu (if there is one).
```
Perhaps this should be a responsibility of maintainers that we could put in the maintainer notes?

Author: I think that only users with triage access can do this, as well as adding labels to issues and adding them to the project board. https://docs.github.com/en/organizations/managing-user-access-to-your-organizations-repositories/managing-repository-roles/repository-roles-for-an-organization#permissions-for-each-role

We should test this. I can ask someone at my uni to try on Friday (making a new account ourselves won't work since we disallowed completely new users to interact with the repos :) )

Reviewer: Ah yes, good point.

Reviewer: Please could you add a comment in the maintainer notes issue to mention this should be included?

* If you are happy to take on a particular Issue, add yourself to the list of assignees using the right-hand panel.

> * Every Pull Request should have an associated Issue. (Think of Issues like a ticketing system.)

Reviewer: ```suggestion
* In general, every Pull Request should have an associated Issue. (Think of Issues like a ticketing system.) An exception is for trivial changes, such as typo fixes. In these cases, feel free to open a PR without a linked issue and add it to the project board.
```

Author: We could create an issue template to explain these things directly there when someone opens an issue. I think that's simpler for everyone and it's probably too optimistic to expect people would read the full instructions in the wiki :)

Reviewer: That sounds like a good idea. I'm generally in support of issue templates, although a note of caution that when issue/PR templates get too complex, people tend to just ignore them. So we need to keep any templates simple.

Author: Opened https://github.com/mesh-adaptation/docs/issues/106

## Pull Requests

When you are ready to start work on an Issue, the recommended workflow is as follows.
Note that if you are not a member of the `mesh-adaptation` organisation then you will have to fork the repository that you wish to contribute to.
1. *Command line:* Check out the `main` branch (or whichever other branch you wish to start work from the head of).
2. *Command line:* Run `git pull` to make sure you have the latest changes.
3. *Command line:* Create a new branch with `git checkout -b XX_<branch_name>`, where `XX` is the Issue number and the branch name uses [snake_case](https://en.wikipedia.org/wiki/Snake_case), e.g., `3_blind_mice`.

> 4. *Command line:* When you commit a change, it can be beneficial to include a link the the issue in the commit message. Do this with `#XX`, where again `XX` is the Issue number<sup>1</sup>. This will provide updates to the Issue on progress that has been made (in the next step). Note: formatting checks may cause your commits to fail on the first attempt - see [below](#formatting) for details.

Reviewer: I wonder if we should drop this. Given that we're always using squash merge, the commit history goes away and links to these references can end up pointing to nothing.

Author: The commit history still stays in the PR, right? So I think it will always still point to that commit. But I'm up for dropping this for other reasons: 
- I think it's not necessary since we have a philosophy of 1 PR per issue, so the issue history gets unnecessary cluttered I think
- I think we're asking too much of people to change their commit message comment from # to something else, or to always use `git commit --cleanup=whitespace` to ignore #

How about we enforce this rule for commit messages of the squashed commit?

Author: And perhaps we could adopt the [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/) approach

Reviewer: > How about we enforce this rule for commit messages of the squashed commit? Then only maintainers actually deal with this since others cannot merge PRs

Agree, let's drop it and switch to this.

Reviewer: > And perhaps we could adopt the [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/) approach

Feel free to add it as a recommendation but it's probably too much to require it.

Author: I dropped this in https://github.com/mesh-adaptation/docs/pull/102/commits/d32e592f6edc306072649f73f86203b60c16591c but I did not add the conventional commits recommendation, if you're okay with that. I think it doesn't make sense to only partially enforce it

4. *Command line* (optional, but recommended): Check out the branch associated with the PR and run tests locally, for example any that have been mentioned specifically in the PR, any new demos introduced, and/or any other cases that you might be concerned about.
5. *GitHub*: Once you are satisfied that you have made a decision about the PR, return to the "Files changed" tab and click the green "Review changes" button on the right-hand side. Choose to simply leave a comment (no conclusion yet), approve, or request changes for the PR and leave a summary comment justifying this choice.
6. *GitHub*: If you did not approve the changes then there may be an iterative back-and-forth between the developer(s) and reviewer(s). In each iteration, follow the process outlined in points 1-5 above.
7. *GitHub*: Upon approving the PR, leave the author to click the green "Merge pull request" button, as they might notice something they forgot at the last minute.

## Testing

All Mesh Adaptation codes have a subdirectory called `test` containing dedicated test suites with a mixture of [unit tests](#unit-tests) and [integration tests](#integration-tests).

> Goalie also has an additional `test_adjoint` subdirectory, specifically for tests of functionality making use of adjoint methods where annotation is turned on.

Reviewer: ```suggestion
Goalie also has an additional `test/adjoint` subdirectory, specifically for tests of functionality making use of adjoint methods where annotation is turned on.
```

These will be automatically installed and set up with the `make install` command.
Upon making any commit to a Mesh Adaptation module, the pre-commit hook will be invoked and will use Ruff to format the code.
If the pre-commit made any additional formatting changes then the commit will fail and you will need to retry.
In some cases, there will still be formatting issues that you will need to fix manually.

## Docker build

The [Mesh Adaptation Docs repository](https://github.com/mesh-adaptation/mesh-adaptation-docs) contains the configuration files for the bespoke Firedrake Docker image mentioned on the [installation instructions page](./Installation-instructions).
This Docker image is based off the ["Firedrake vanilla"](https://github.com/firedrakeproject/firedrake/blob/master/docker/Dockerfile.vanilla) one.
It starts from that Docker image and makes some modifications:
* Switch to the [`jwallwork23/firedrake-parmmg`](https://github.com/firedrakeproject/petsc/tree/jwallwork23/firedrake-parmmg) PETSc branch;
* Reconfigure PETSc, installing Mmg, ParMmg, and their dependencies;
* Rebuild PETSc and its other dependencies;
* Rebuild Firedrake and its other dependencies;

> * Install Animate, Goalie, Movement, and their other dependencies.

Reviewer: I think this needs updating? For starters, we're no longer using a custom PETSc branch.

Author: I updated this in https://github.com/mesh-adaptation/docs/pull/102/commits/370f30b398b8e283415b6f319185c19456466f7c without mentioning the UM2N image. UM2N is completely missing from the Wiki so I'll open a new issue for that
